### PR TITLE
fix: save revealed accounts regardless of permissions granted

### PR DESCRIPTION
**Pull request by MishkaRogachev**
For https://github.com/status-im/status-desktop/pull/10967

Important changes:
- [x] save revealed accounts regardless of permissions granted

This PR does not contain checks for provided addresses, this should be fixed in https://github.com/status-im/status-go/pull/3580 after release

### CODE REVIEW

**Review by jrainville — APPROVED** [review]

File: protocol/communities_messenger_test.go [hunk 1]
```diff
@@ -126,33 +175,52 @@ func (s *MessengerCommunitiesSuite) newMessengerWithOptions(shh types.Waku, priv
...
+	// add wallet account with keypair
+	kp := accounts.GetProfileKeypairForTest(false, true, false)
+	kp.Accounts[0].Address = types.HexToAddress(walletAddress)
+	err = m.settings.SaveOrUpdateKeypair(kp)
+	s.Require().NoError(err)
+
+	walletAccounts, err := m.settings.GetAccounts()
+	s.Require().NoError(err)
+	s.Require().Len(walletAccounts, 1)
+	//	s.Require().Equal(walletAccounts[0].Type, accounts.AccountTypeGenerated)
```
Comment (jrainville) [inline protocol/communities_messenger_test.go]: Leftover?

**Review by jrainville — APPROVED** [review]

File: protocol/communities/manager.go [hunk 2]
```diff
@@ -1783,6 +1783,38 @@ func (m *Manager) HandleCommunityRequestToJoin(signer *ecdsa.PublicKey, request
 		return nil, err
 	}
 
+	// Save revealed addresses + signatures so they can later be added
+	// to the community member list when the request is accepted
+	if len(request.RevealedAccounts) > 0 {
+		err = m.persistence.SaveRequestToJoinRevealedAddresses(requestToJoin)
```
Comment (mprakhov) [inline protocol/communities/manager.go]: I think this should be at the end, only when account was verified

Comment (MishkaRogachev) [reply protocol/communities/manager.go]: fixed

File: protocol/communities/manager.go [hunk 3]
```diff
@@ -1845,7 +1845,38 @@ func (m *Manager) HandleCommunityRequestToJoin(signer *ecdsa.PublicKey, request
...
-		return requestToJoin, nil
+	}
+
+	// Save revealed addresses + signatures so they can later be added
+	// to the community member list when the request is accepted
+	if len(request.RevealedAccounts) > 0 {
+		err = m.persistence.SaveRequestToJoinRevealedAddresses(requestToJoin)
+		if err != nil {
+			return nil, err
+		}
```
Comment (0x-r4bbit) [inline protocol/communities/manager.go]: This looks a big backwards to me. Shouldn't we only save the revealed addresses, **after** we've verified the signatures?

File: protocol/communities_messenger_test.go [hunk 4]
```diff
@@ -3585,3 +3654,27 @@ func (s *MessengerCommunitiesSuite) TestHandleImport() {
...
+	s.advertiseCommunityTo(community, s.bob)
+
+	s.joinCommunity(community, s.alice, AlicePassword)
+	s.joinCommunity(community, s.bob, BobPassword)
+
+	joinedCommunities, err := s.admin.communitiesManager.Joined()
+	s.Require().NoError(err)
+	s.Require().Len(joinedCommunities, 1)
+	community = joinedCommunities[0]
+	s.Require().NotNil(community)
```
Comment (0x-r4bbit) [inline protocol/communities_messenger_test.go]: Another way to do this would be to do `s.admin.GetCommunityByID(community.ID())`

File: protocol/communities_messenger_test.go [hunk 5]
```diff
@@ -3585,3 +3654,27 @@ func (s *MessengerCommunitiesSuite) TestHandleImport() {
...
+
+	s.Require().Equal(3, community.MembersCount())
+
+	for pubKey, member := range community.Members() {
+		if pubKey != common.PubkeyToHex(&s.admin.identity.PublicKey) {
+			s.Require().Len(member.RevealedAccounts, 1)
+			s.Require().Equal(member.RevealedAccounts[0].Address, walletAddress)
+		}
```
Comment (0x-r4bbit) [inline protocol/communities_messenger_test.go:3675]: Since we're iterating over members here, do we expect both, `bob` and `alice` to reveal the exact same `walletAddress`?

File: protocol/communities/manager.go (hunk 3 above)
Comment (MishkaRogachev) [reply protocol/communities/manager.go]: fixed

File: protocol/communities_messenger_test.go (hunk 4 above)
Comment (MishkaRogachev) [reply protocol/communities_messenger_test.go]: Yes, i just what to check here we are actually joined

Comment (0x-r4bbit) [reply protocol/communities_messenger_test.go]: `joined()` will return the communities you're joined. But admin has created that community, so it will always be joined.

File: protocol/communities_messenger_test.go (hunk 5 above)
Comment (MishkaRogachev) [reply protocol/communities_messenger_test.go:3675]: Yes, it is done consciously. The problem is in `communities/manager.go:1860`:
```
	recovered, err := m.accountsManager.Recover(recoverParams)
	if err != nil {
		return nil, err
	}
	if recovered.Hex() != revealedAccount.Address {
```
We can have different wallet addresses in the test, but that requires forging that address every time we ask to join with a valid wallet address, and that's a pretty painful change. Moreover, tests becoming *more* fragile, because we are trying to fake behaviour of an object, that is far away from issue's scope

Comment (MishkaRogachev) [reply protocol/communities_messenger_test.go:3675]: The right way i see is to change `AccountManager::Recover` api to embed verification inside, but i would wait for @saledjenic to check it with him

File: protocol/communities_messenger_test.go (hunk 4 above)
Comment (MishkaRogachev) [reply protocol/communities_messenger_test.go]: fixed

**Review by mprakhov — APPROVED** [review]

**Review by 0x-r4bbit — APPROVED** [review]

File: account/accounts.go [hunk 6]
```diff
@@ -62,7 +62,14 @@ type RecoverParams struct {
 	Signature string `json:"signature"`
 }
 
-// Manager represents account manager interface.
+// Interface represents account manager interface
+type Interface interface {
+	GetVerifiedWalletAccount(db *accounts.Database, address, password string) (*SelectedExtKey, error)
+	Sign(rpcParams SignParams, verifiedAccount *SelectedExtKey) (result types.HexBytes, err error)
+	Recover(rpcParams RecoverParams) (addr types.Address, err error)
+}
```
Comment (Samyoul) [inline account/accounts.go:70]: This interface name doesn't describe the kind of interface, it is already declared as an interface type so this information doesn't add anything to the code.

If we look at `stdlib` interfaces they have names like `JSONMarshaller`, `Reader`, `Writer` and `Stringer`. The convention is to name the interfaces after the action they perform.

In this case the interfaces `Get`s a `VerifiedWalletAccount`, `Sign`s and `Recover`s, the name would be something that reflects this functionality. Example `AccountManager`, `KeyManager` or `KeyHandler`

File: protocol/communities_messenger_test.go [hunk 7]
```diff
@@ -63,9 +108,9 @@ func (s *MessengerCommunitiesSuite) SetupTest() {
 	s.shh = gethbridge.NewGethWakuWrapper(shh)
 	s.Require().NoError(shh.Start())
 
-	s.admin = s.newMessenger()
-	s.bob = s.newMessenger()
-	s.alice = s.newMessenger()
+	s.admin = s.newMessenger(AdminPassword)
+	s.bob = s.newMessenger(BobPassword)
+	s.alice = s.newMessenger(AlicePassword)
```
Comment (Samyoul) [inline protocol/communities_messenger_test.go:113]: For the tests do we need to define the password at this level? Could passwords be an empty string or the same password used all messengers? 

Function signatures have been changed to accommodate the concept of password, but the password is not the testable element.

**Review by Samyoul — COMMENTED** [review]
@MishkaRogachev Thank you for this PR, I appreciate it is already been merged, but perhaps you could give some thought to the points I've raised here. Thank you again 😄

File: account/accounts.go (hunk 6 above)
Comment (MishkaRogachev) [reply account/accounts.go:70]: Sure, i was confused by name `Manager`, so i  thought `accounts.Interface` and `accounts.Manager` is a good pair.
And i created an issue for fixing stuff after this PR: https://github.com/status-im/status-go/issues/3625

File: protocol/communities_messenger_test.go (hunk 7 above)
Comment (MishkaRogachev) [reply protocol/communities_messenger_test.go:113]: The problem is that join to community requires password, and if its not match the behaviour should be different.  An i suppose we should test that. If i set same password for each community member, test may not cover some edge case.
